crates/wasmtime/src/runtime/store.rs

```diff
@@ -2116,6 +2124,27 @@ unsafe impl<T> crate::runtime::vm::VMStore for StoreInner<T> {
                         self.async_yield_impl()?;
                         delta
                     }
+                    #[cfg(feature = "async")]
+                    UpdateDeadline::YieldCustom(delta, future) => {
+                        assert!(
+                            self.async_support(),
+                            "cannot use `UpdateDeadline::YieldCustom` without enabling async support in the config"
+                        );
+
+                        // When control returns, we have a `Result<()>` passed
+                        // in from the host fiber. If this finished successfully then
+                        // we were resumed normally via a `poll`, so keep going.  If
+                        // the future was dropped while we were yielded, then we need
+                        // to clean up this fiber. Do so by raising a trap which will
+                        // abort all wasm and get caught on the other side to clean
+                        // things up.
+                        unsafe {
+                            self.async_cx()
+                                .expect("attempted to pull async context during shutdown")
+                                .block_on(::core::pin::Pin::new_unchecked(future))?
```

Comment:
I realize this was probably copied from `async_yield_impl` but nowadays I think we can do `let mut future = core::pin::pin!(future);` to safely creat `Pin<Thing>` and then with some `future.as_mut()` or such invocations I think it can safely create `Pin<&mut T>` to pass here

Reply:
In this particular case, I think we can just change the type signature of the future to be a `Pin<Box<dyn Future ...>>>` to avoid having to do anything extra to mark the memory as pinned.